qcodes/instrument_drivers/devices.py

```diff
@@ -1,6 +1,190 @@
 from typing import Union, cast
+from operator import xor
+import enum
+import warnings
 
-from qcodes import Parameter, Instrument
+from qcodes import Parameter, ManualParameter, Instrument
+
+
+class ParameterScaler(Parameter):
+    """
+    Parameter Scaler
+
+    To be used when you use a physical voltage divider or an amplifier to set
+    or get a quantity.
+
+    Initialize the parameter by passing the parameter to be measured/set
+    and the value of the division OR the gain.
+
+    The scaling value can be either a scalar value or a Qcodes Parameter.
+
+    The parameter scaler acts a your original parameter, but will set the right
+    value, and store the gain/division in the metadata.
+
+    Examples:
+        Resistive voltage divider
+        >>> vd = ParameterScaler(dac.chan0, division = 10)
+
+        Voltage multiplier
+        >>> vb = ParameterScaler(dac.chan0, gain = 30, name = 'Vb')
+
+        Transimpedance amplifier
+        >>> Id = ParameterScaler(multimeter.amplitude, division = 1e6, name = 'Id', unit = 'A')
+
+    Args:
+        output: Physical Parameter that need conversion
+        division: the division value
+        gain: the gain value
+        label: label of this parameter, by default uses 'output' label
+            but attaches _amplified or _attenuated depending if gain
+            or division has been specified
+        name: name of this parameter, by default uses 'output' name
+            but attaches _amplified or _attenuated depending if gain
+            or division has been specified
+        unit: resulting unit. It uses the one of 'output' by default
+    """
+
+    class Role(enum.Enum):
+        GAIN = enum.auto()
+        DIVISION = enum.auto()
+
+
+    def __init__(self,
+                 output: Parameter,
+                 division: Union[int, float, Parameter] = None,
+                 gain: Union[int, float, Parameter] = None,
+                 name: str=None,
+                 label: str=None,
+                 unit: str=None) -> None:
+        self.wrapped_parameter = output
+        self.wrapped_instrument = getattr(output, "_instrument", None)
+
+        # Set the role, either as divider or amplifier
+        # Raise an error if nothing is specified
+        is_divider = division is not None
+        is_amplifier = gain is not None
+
+        if not xor(is_divider, is_amplifier):
+            raise ValueError('Provide only division OR gain')
```

Comment:
Why not allow both? Have both default to `1` instead of `None`, drop `self.Role`, in the constructor set `self._multiplier = gain / division`, and have `get_raw` return `self.wrapped_parameter() * self._multiplier()`? I'm probably missing something, but that seems simpler in both usage and implementation.

Reply:
I don't see an use case for having both a multiplier and a divider. Currently they are just defined as the mutual reciprocal (divider = 1/gain or gain = 1/divider) as user convenience. For example, for a a resistive divider, you can say that is divided by 1000 or that it has a gain of 0.001, and it's totally equivalent

Reply:
I guess the root of my confusion is that it doesn't seem like the `Role` enum is necessary at all. If they call `division` then just divide`_multiplier` by the given factor.

You have this:
```python
    class Role(enum.Enum):
        GAIN = enum.auto()
        DIVISION = enum.auto()


    def __init__(self,
                 output: Parameter,
                 division: Union[int, float, Parameter] = None,
                 gain: Union[int, float, Parameter] = None,
                 name: str=None,
                 label: str=None,
                 unit: str=None) -> None:

        # (...)

        # Set the role, either as divider or amplifier
        # Raise an error if nothing is specified
        is_divider = division is not None
        is_amplifier = gain is not None

        if not xor(is_divider, is_amplifier):
            raise ValueError('Provide only division OR gain')

        if is_divider:
            self.role = ParameterScaler.Role.DIVISION
            self._multiplier = division
        elif is_amplifier:
            self.role = ParameterScaler.Role.GAIN
            self._multiplier = gain

        # (...)

    # Division of the scaler
    @property
    def division(self):
        if self.role == ParameterScaler.Role.DIVISION:
            return self._multiplier()
        elif self.role == ParameterScaler.Role.GAIN:
            return 1 / self._multiplier()

    @division.setter
    def division(self, division: Union[int, float, Parameter]):
        self.role = ParameterScaler.Role.DIVISION
        self._multiplier = division

    # Getter and setter for the real value
    def get_raw(self) -> Union[int, float]:
        """
        Returns:
            number: value at which was set at the sample
        """
        if self.role == ParameterScaler.Role.GAIN:
            value = self._wrapped_parameter() * self._multiplier()
        elif self.role == ParameterScaler.Role.DIVISION:
            value = self._wrapped_parameter() / self._multiplier()

        self._save_val(value)
        return value
```
Could you do this instead? To me at least it is much simpler.
```python
    def __init__(self,
                 output: Parameter,
                 division: Union[int, float, Parameter] = 1,
                 gain: Union[int, float, Parameter] = 1,
                 name: str=None,
                 label: str=None,
                 unit: str=None) -> None:

        # (...)

        self._multiplier = gain / division

        # (...)

    # Division of the scaler
    @property
    def division(self):
        return self._multiplier()

    @division.setter
    def division(self, division: Union[int, float, Parameter]):
        self._multiplier = 1/division

    # Getter and setter for the real value
    def get_raw(self) -> Union[int, float]:
        """
        Returns:
            number: value at which was set at the sample
        """
        value = self._wrapped_parameter() * self._multiplier()

        self._save_val(value)
        return value
```

Not a big deal though, feel free to ignore :)

Reply:
I like the idea of @spxtr that the `Role` is not needed, but the `self._multiplier = gain / division` part seems confusing. I'd suggest that one can initialize a ParameterScaler via `gain` *OR* `division` kwarg (as it is now), but instead of having a `Role`, either `gain` or `division` are converted to `multiplier`. I mean smth like this:
```
    def __init__(self,
                 output: Parameter,
                 division: Union[int, float, Parameter] = 1,
                 gain: Union[int, float, Parameter] = 1,
                 name: str=None,
                 label: str=None,
                 unit: str=None) -> None:

        # (...)

        
        if not xor(gain is None, division is None):
            raise KeyError('Either gain or division should be supplied')

        if gain:  # gain is defined as scaled_value/wrapped_value
            self.gain = gain  # magic happens in the property setter
        if division:  # division is defined as 1/(scaled_value/wrapped_value)
            self.division = division  # magic happens in the property setter

        # (...)

    # Division of the scaler
    @property
    def division(self):
        return 1/self._multiplier()

    @division.setter
    def division(self, division: Union[int, float, Parameter]):
        self._multiplier.set(1/division)

    # Gain of the scaler
    @property
    def gain(self):
        return self._multiplier()

    @gain.setter
    def gain(self, gain: Union[int, float, Parameter]):
        self._multiplier.set(gain)

    # Getter for the real value
    def get_raw(self) -> Union[int, float]:
        """
        Returns:
            number: value at which was set at the sample
        """
        value = self._wrapped_parameter() * self._multiplier()

        self._save_val(value)
        return value
```
This eliminates the need for the role (testing is easier, no "factory" requests from me :) ).

@YakBizzarro what do you think?

Reply:
Role is needed in the case when the multiplier is not a simple scalar, but a Qcodes parameter (look at the last example). In this case 1/division should be computed every time that we set a value, or (like a do), perform a multiplication, depending on the role value

Reply:
Now I get it: if the multiplier is a qcodes parameter, then the ScaledParameter needs to know what the role of the multiplier is in order to perform the right operation (multiplication or division)...

ok, the number of responsibilities/blows/whistles of the ScaledParameter is exploding.

Could you list the use cases that this class is intended to support? Is the following list complete ? :
* create a new parameter that has a *fixed and unchangeable* scaling *factor* (both <1 or >1) for another parameter
* create a new parameter that *multiplies* another parameter with a *third parameter*
* create a new parameter that *divides* another parameter by a *third parameter*
If so, then I see at least 3 different classes (and possibly a couple of extra private ones to "modularize" the implementation).

Reply:
Right. Just in the first case, the scaling factor can be changed manually by the user using the setter of division/gain. Useful for IVVI modules with manual gain/division

Reply:
The latter cases are intended for instruments where the gain is selectable by the computer or when is function of an external value